### ISSUE TITLE
Unstable onPress reference with interaction tracking enabled

### ISSUE THREAD

**Issue by themichaellai** [DataDog/dd-sdk-reactnative#138]
Hey! We're observing that components that accept an `onPress` prop are getting different `onPress` references on each render, despite passing what should be the same reference. This is causing components with an `onPress` callback that we've `React.memo`ed to re-render when they shouldn't be.

**Steps to reproduce the issue:**
This should affect components that have an `onPress` function prop that are `React.memo`ed — we've reproduced it in our app with a component like:
```typescript
import React from 'react';
import { Text } from 'react-native';

interface Props {
  text: string;
  onPress: () => void;
}
const FooBar: React.FC<Props> = ({ text, onPress }) => {
  console.log('Rendering FooBar');
  return <Text>{text}</Text>;
};

const propsAreEqual = (prevProps: Props, nextProps: Props) => {
  console.log('--- props are equal...')
  console.log(`onPress: `, prevProps.onPress === nextProps.onPress);
  console.log(`text: ${prevProps.text === nextProps.text}`)
  return (
    prevProps.text === nextProps.text &&
    prevProps.onPress === nextProps.onPress
  );
};
const FooBarMemo = React.memo(FooBar, propsAreEqual);

const ParentOfFooBar = () => {
  const foobarCallback = useCallback(() => {
    console.log('foobarCallback');
  }, []);
  // FooBarMemo re-renders anytime ParentOfFooBar re-renders
  return (
    <FooBarMemo onPress={foobarCallback} text="hello world foobar" />
  );
};
```
with SDK configuration that looks like this:
```typescript
const config = new DdSdkReactNativeConfiguration(
  '...',
  environment,
  '...',
  true, // track User interactions
  true, // track XHR Resources
  true, // track Errors
);
```

Disabling interaction tracking allows the component to not re-render as expected.

I believe this is due to the `onPress` function being replaced every render here: https://github.com/DataDog/dd-sdk-reactnative/blob/bd8cc44bd4512117328bad546a2d42b1599f3e0c/packages/core/src/rum/instrumentation/DdRumUserInteractionTracking.tsx#L40-L43

We're able to side-step this by disabling interaction tracking.

We're using `@datadog/mobile-react-native@1.0.0-rc3`.

**Comment by 0xnm** [comment]
Hello @themichaellai! Thanks for reporting that, we will take a look.

**Comment by 0xnm** [comment]
Hello @themichaellai! First of all, thanks a lot for the detailed issue report.

It seems we won't be able to fix it reliably without a breaking change in the API (changing mechanics of automated interactions tracking). The way it works, it needs to wrap every `onPress` function, so once new element is created we need to create a new wrapper as well. We could store already created wrappers for the `onPress` functions we already seen in the `WeakMap` to support memoization, but this is not a 100% reliable solution.

So, for now, I would suggest just to provide a custom comparison function in the `React.memo` call. By default, `React.memo` is using shallow comparison which is declared [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/Libraries/Renderer/implementations/ReactFabric-prod.js#L2067-L2090) (there are different renderer implementations, but the part of shallow comparison for memoized components is pretty much the same between them), with usage example [here](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/Libraries/Renderer/implementations/ReactFabric-prod.js#L4180). Considering that, it is possible to give the same function to the `React.memo` call with a small modification which will just drop `onPress` prop from `keys` arrays for the comparison.

Sorry for the inconvenience, we will revisit interactions tracking in the later major updates.

**Comment by themichaellai** [comment]
Thanks for looking into this @0xnm! We'll try using the `React.memo` comparison function.

**Comment by louiszawadzki** [comment]
Hi @themichaellai , this issue has been solved in version 1.0.0-rc7!

**Comment by themichaellai** [comment]
Thanks — this has been working well for us !
